run/test/specs/utils/screenshot_helper.ts

```diff
@@ -0,0 +1,259 @@
+// run/test/specs/utils/screenshot-helper.ts
+import { DeviceWrapper } from '../../../types/DeviceWrapper';
+import type { TestInfo } from '@playwright/test';
+import * as fs from 'fs';
+import * as path from 'path';
+import sharp from 'sharp';
+import { SupportedPlatformsType } from './open_app';
+
+// Screenshot context type
+type ScreenshotContext = {
+  devices: DeviceWrapper[];
+  testInfo: TestInfo;
+  platform: SupportedPlatformsType;
+};
+
+// Global registry to track devices for screenshot capture
+const deviceRegistry = new Map<string, ScreenshotContext>();
+
+// Register devices for a test
+export function registerDevicesForTest(
+  testInfo: TestInfo,
+  devices: DeviceWrapper[],
+  platform: SupportedPlatformsType
+) {
+  const testId = `${testInfo.testId}-${testInfo.repeatEachIndex}`;
+  deviceRegistry.set(testId, { devices, testInfo, platform });
+}
+
+// Unregister devices after test
+export function unregisterDevicesForTest(testInfo: TestInfo) {
+  const testId = `${testInfo.testId}-${testInfo.repeatEachIndex}`;
+  deviceRegistry.delete(testId);
+}
+// Add device labels to screenshots (e.g. "Device: alice1")
+async function addDeviceLabel(screenshot: Buffer, device: DeviceWrapper): Promise<Buffer> {
+  const metadata = await sharp(screenshot).metadata();
+  const width = metadata.width;
+  const deviceName = device.getDeviceIdentity();
+
+  // Create transparent label overlay
+  const labelHeight = 60;
+  const fontSize = 28;
+  const label = Buffer.from(`
+    <svg width="${width}" height="${labelHeight}">
+      <rect x="0" y="0" width="${width}" height="${labelHeight}"
+            fill="black" fill-opacity="0"/>
+      <text x="${width / 2}" y="${labelHeight / 2 + fontSize / 3}" 
+            font-family="-apple-system, Arial, sans-serif" 
+            font-size="${fontSize}" 
+            font-weight="bold"
+            fill="white" 
+            text-anchor="middle">
+         Device: ${deviceName}
+      </text>
+    </svg>
+  `);
+
+  // Composite label over screenshot
+  return sharp(screenshot)
+    .composite([
+      {
+        input: label,
+        top: 0,
+        left: 0,
+        blend: 'over',
+      },
+    ])
+    .png()
+    .toBuffer();
+}
+
+async function createSideBySideComposite(left: Buffer, right: Buffer): Promise<Buffer> {
+  try {
+    // Get dimensions of first image
+    const metadata = await sharp(left).metadata();
+    const width = metadata.width;
```

Comment:
search for metadata in all files, but this can be cleaned up with 
`const {width} = await sharp(left).metadata();`